### PR TITLE
Admin API Auth

### PR DESCRIPTION
**Pull request by landonpierce**
Admin API now requires a bearer token issued from azure ad b2c to communicate with it. 

Because of b2c limitations, communication between the admin api and permissions api is secured via mutual tls certificate auth. The admin api will fetch the certificate from a keyvault and will use it when it issues a request to the permissions api, which will verify the thumbprint matches what it expects before returning any data. 

Still To do on the admin api:
- Implement data authorization based on signed in user
- Figure out what scopes we want to authorize for each route
- Add in options for turning back on OBO flow if user wanted to use AAD instead of b2c

### CODE REVIEW

File: src/Saas.Admin/Saas.Admin.Service/Controllers/TenantsController.cs [hunk 1]
```diff
@@ -1,10 +1,14 @@
 ﻿#nullable disable
 
+using Microsoft.AspNetCore.Authorization;
```
Comment (Codebytes) [inline src/Saas.Admin/Saas.Admin.Service/Controllers/TenantsController.cs:3]: move to global usings

File: src/Saas.Admin/Saas.Admin.Service/Controllers/TenantsController.cs [hunk 2]
```diff
@@ -208,7 +212,7 @@ public async Task<IActionResult> DeleteTenant(Guid tenantId)
     [ProducesResponseType(StatusCodes.Status200OK)]
     [ProducesResponseType(StatusCodes.Status401Unauthorized)]
     [ProducesResponseType(StatusCodes.Status404NotFound)]
-    public async Task<ActionResult<IEnumerable<string>>> GetTenantUsers(Guid tenantId)
+    public async Task<ActionResult<IEnumerable<string>>> GetTenantUsers(string tenantId)
```
Comment (Codebytes) [inline src/Saas.Admin/Saas.Admin.Service/Controllers/TenantsController.cs:215]: Why the change? for model binding?

File: src/Saas.Admin/Saas.Admin.Service/Program.cs [hunk 3]
```diff
@@ -1,21 +1,54 @@
 using Saas.Admin.Service.Data;
+using Microsoft.IdentityModel.Logging;
```
Comment (Codebytes) [inline src/Saas.Admin/Saas.Admin.Service/Program.cs:2]: move to global usings

File: src/Saas.Admin/Saas.Admin.Service/Program.cs [hunk 4]
```diff
@@ -1,21 +1,54 @@
...
+builder.Services.AddAuthorization(options => { });
+
 
 builder.Services.AddControllers();
 
 builder.Services.AddScoped<ITenantService, TenantService>();
 builder.Services.AddScoped<IPermissionService, PermissionService>();
 
+// Use azure keyvault SDK to download certificate to be used to authenticate with permissions api
+CertificateClient certificateClient = new CertificateClient(new Uri(builder.Configuration["KeyVault:Url"]), new DefaultAzureCredential());
```
Comment (Codebytes) [inline src/Saas.Admin/Saas.Admin.Service/Program.cs:29]: nice, glad we added kv

File: src/Saas.Admin/Saas.Admin.Service/Program.cs [hunk 5]
```diff
@@ -35,6 +68,9 @@
 {
     app.UseSwagger();
     app.UseSwaggerUI();
+
+    // Show identity debug information during development. This should be kept off in production to comply with GDPR standards
+    IdentityModelEventSource.ShowPII = true;
```
Comment (Codebytes) [inline src/Saas.Admin/Saas.Admin.Service/Program.cs:73]: we need to be careful with this setting

File: src/Saas.Admin/Saas.Admin.Service/Services/IPermissionService.cs [hunk 6]
```diff
@@ -2,9 +2,9 @@
 
 public interface IPermissionService
 {
-    Task<IEnumerable<string>> GetTenantUsersAsync(Guid tenantId);
-    Task<IEnumerable<string>> GetUserPermissionsForTenantAsync(Guid tenantId, string userId);
-    Task AddUserPermissionsToTenantAsyc(Guid tenantId, string userId, string[] permissions);
-    Task RemoveUserPermissionsFromTenantAsync(Guid tenantId, string userId, string[] permissions);
-    Task<IEnumerable<Guid>> GetTenantsForUserAsync(string userId, string? filter);
+    Task<IEnumerable<string>> GetTenantUsersAsync(string tenantId);
```
Comment (Codebytes) [inline src/Saas.Admin/Saas.Admin.Service/Services/IPermissionService.cs:5]: they should really be guids on the interface, right?

File: src/Saas.Admin/Saas.Admin.Service/appsettings.json [hunk 7]
```diff
@@ -1,12 +1,11 @@
 {
-  "AzureAd": {
-    "Instance": "https://login.microsoftonline.com/",
-    "Domain": "saaskit.onmicrosoft.com",
-    "TenantId": "9d68116c-e8ee-4ee6-9ab5-5c7f11b450bb",
-    "ClientId": "f5fa7382-59b3-4b9f-b996-2a48e3b20ce7",
-    "Scopes": "access_as_user",
-    "CallbackPath": "/signin-oidc",
-    "SignedOutCallbackPath": "/signout-callback-oidc"
+  "AzureAdB2C": {
```
Comment (Codebytes) [inline src/Saas.Admin/Saas.Admin.Service/appsettings.json:2]: these should be empty strings

File: src/Saas.Admin/Saas.Admin.Service/Program.cs (hunk 5 above)
Comment (landonpierce) [reply src/Saas.Admin/Saas.Admin.Service/Program.cs:73]: Thanks for the catch. I have only turned on in development on the permissions API. I'll add a check here too.

File: src/Saas.Admin/Saas.Admin.Service/Controllers/TenantsController.cs (hunk 2 above)
Comment (landonpierce) [reply src/Saas.Admin/Saas.Admin.Service/Controllers/TenantsController.cs:215]: @asanjabi and I went back and forth on this a bit. We ultimately decided it was better to use strings across the board for IDs incase the user wanted to use something other than GUIDs for tenantIDs.

File: src/Saas.Admin/Saas.Admin.Service/Services/IPermissionService.cs (hunk 6 above)
Comment (landonpierce) [reply src/Saas.Admin/Saas.Admin.Service/Services/IPermissionService.cs:5]: See comment above

**Review by Codebytes — APPROVED** [review]
i understand the string<>guid discussion.
rest looks fine

File: src/Saas.Admin/Saas.Admin.Service/Program.cs [hunk 8]
```diff
@@ -1,21 +1,54 @@
...
 var builder = WebApplication.CreateBuilder(args);
 
 builder.Services.AddDbContext<TenantsContext>(options =>
     options.UseSqlServer(builder.Configuration.GetConnectionString("TenantsContext")));
 
-// Add services to the container.
-builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
-    .AddMicrosoftIdentityWebApi(builder.Configuration.GetSection("AzureAd")).EnableTokenAcquisitionToCallDownstreamApi()
-            .AddDownstreamWebApi("DownstreamApi", builder.Configuration.GetSection("DownstreamApi"))
-            .AddInMemoryTokenCaches();
```
Comment (asanjabi) [inline src/Saas.Admin/Saas.Admin.Service/Program.cs:12]: 👿

File: src/Saas.Admin/Saas.Admin.Service/appsettings.json (hunk 7 above)
Comment (asanjabi) [reply src/Saas.Admin/Saas.Admin.Service/appsettings.json:2]: Nothing secret here really, agree for final version though, to force users to create their own.

**Review by asanjabi — APPROVED** [review]
